woocommerce/woocommerce-gutenberg-products-block
# Allow rendering HTML in shipping package names

Fixes #6903

### Note

Currently, it's not possible to render HTML elements when adding a filter for the shipping package name. This PR aims to enable render HTML when using the shipping package name filter.

Based on @tarhi-saad's [review](https://github.com/woocommerce/woocommerce-blocks/pull/7147#discussion_r977787322), this PR also adds the reusable util `sanitizeHTML`.

### Screenshots

#### Cart block

<table>
<tr>
<td valign="top">Before:
<br><br>

<img width="410" alt="Screenshot 2022-09-14 at 12 11 09" src="https://user-images.githubusercontent.com/3323310/190064821-e74cd262-88c6-444f-a54b-864e981068e5.png">
</td>
<td valign="top">After:
<br><br>

<img width="414" alt="Screenshot 2022-09-14 at 12 10 23" src="https://user-images.githubusercontent.com/3323310/190064731-3761de4f-ecf9-4e19-9537-edc7e752c298.png">
</td>
</tr>
</table>

#### Checkout block

<table>

<tr>
<td valign="top">Before:
<br><br>

<img width="732" alt="Screenshot 2022-09-14 at 12 11 46" src="https://user-images.githubusercontent.com/3323310/190064882-05903fb9-3529-4692-ba33-c36900503fb5.png">
</td>
<td valign="top">After:
<br><br>

<img width="731" alt="Screenshot 2022-09-14 at 12 10 00" src="https://user-images.githubusercontent.com/3323310/190064681-eb5c2658-a9cd-458b-b7ef-0239c62a5e05.png">
</td>
</tr>
</table>

### Testing

#### Test HTML rendering in shipping package names

1. Install the [Multiple Packages for WooCommerce](https://wordpress.org/plugins/multiple-packages-for-woocommerce/) plugin.
2. Go to `/wp-admin/admin.php?page=wc-settings&tab=multiple_packages`, enable the Multiple Packages for Shipping plugin and select `Product (individual)` for the option `Group By`.
3. Install the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin.
4. Add the following code snippet to the site:
```php
add_filter( 'woocommerce_shipping_package_name', 'test_package_name', 1, 3 );
function test_package_name( $title, $index, $package ) {
	return $title . '<br/>test<br/>test2';
}
```
5. Ensure to have one page with the Cart block and one page with the Checkout block.
6. Add two products to the cart.
7. Verify that the page with the Cart block shows the the shipping package name with line breaks instead of showing `Shipping method<br/>test<br/>test2`.
8. Verify that the page with the Checkout block shows the the shipping package name with line breaks instead of showing `Shipping method<br/>test<br/>test2`.
9. Replace the previous code snippet with the following one:
```php
add_filter( 'woocommerce_shipping_package_name', 'test_xss', 1, 3 );
function test_xss( $title, $index, $package ) {
    return $title . '<br><a href="https://woocommerce.com/" target="_blank">WooCommerce.com<img src=x onerror="&#0000106&#0000097&#0000118&#0000097&#0000115&#0000099&#0000114&#0000105&#0000112&#0000116&#0000058&#0000097&#0000108&#0000101&#0000114&#0000116&#0000040&#0000039&#0000088&#0000083&#0000083&#0000039&#0000041"></a>';
}
```
10. Verify that the page with the Cart block and the page with the Checkout block:
  - do not show an XSS alert.
  - do show the package name and the URL [WooCommerce.com](https://woocommerce.com/) in the next line and that the URL opens in a new tab when being clicked.

#### Test reusable sanitise util (only for developer testing)

1. Open `assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx` and change 
```tsx
<div
  className="wc-block-components-shipping-rates-control__package-title"
  dangerouslySetInnerHTML={ sanitizeHTML( packageData.name ) }
/>
```
to
```tsx
<div
  className="wc-block-components-shipping-rates-control__package-title"
  dangerouslySetInnerHTML={ sanitizeHTML( packageData.name, { tags: [ 'a' ] } ) }
/>
```
2. Run `npm run build` to build the extension again.
3. Look up the page with the Cart or the Checkout block from the previous test case.
4. Verify that the the URL [WooCommerce.com](https://woocommerce.com/) is still visible, but that it no longer appears in the next line but in the same line.

### WooCommerce Visibility


### Changelog

> Fix: Render HTML elements on Cart and Checkout blocks when using the `woocommerce_shipping_package_name` filter.
